File: javascript/easy/secondMinimumNodeInABinaryTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var findSecondMinimumValue = function(root) {
    const min = root.val;
    const secondMin = findSecondMin(root, min);
    return (secondMin === Number.MAX_SAFE_INTEGER) ? -1 : secondMin;
};

var findSecondMin = function(root, min) {
    if(root === null) return Number.MAX_SAFE_INTEGER;
    let secondMin = (root.val > min) ? root.val : Number.MAX_SAFE_INTEGER;
    return Math.min(secondMin, findSecondMin(root.left, min), findSecondMin(root.right, min));
};

File: ruby/easy/second_minimum_node_in_a_binary_tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Integer}
def find_second_minimum_value(root)
    second_min = find_second_minimum(root, root.val)
    return (second_min == ((2 ** 31) - 1)) ? -1 : second_min
end

def find_second_minimum(root, min)
    return ((2 ** 31) - 1) if root.nil?
    second_min = (root.val > min) ? root.val : ((2 ** 31) - 1)
    return [second_min, find_second_minimum(root.left, min), find_second_minimum(root.right, min)].min
end
